### ISSUE TITLE
Remove redundant checks for taints and nodeSelectors from strategies.

### ISSUE THREAD

**Issue by RyanDevlin** [kubernetes-sigs/descheduler#574]
**Is your feature request related to a problem? Please describe.**

Currently many features implement their own checks to determine if a pod will end up on the same node after eviction. Some strategies check for taints, others check for nodeSelectors, others check both, and some don't check for this edge case. With the addition of the `nodeFit` feature, checks for taints, nodeSelectors, and whether a node is unschedulable can be turned on and off for almost all strategies. The `nodeFit` feature allows for a consistent approach to this optimization so that all strategies are in line with one another.

**Describe the solution you'd like**

Ideally, every strategy should explicitly *not* check for taints and/or nodeSelectors. If the consideration of these parameters is desired, one could enable the `nodeFit` strategy parameter and achieve the same optimization.

**Describe alternatives you've considered**

None.

**What version of descheduler are you using?**
Any version.

**Additional context**

See the discussion in #559 for more detailed information regarding exactly where edits should be made to the codebase.

**Comment by RyanDevlin** [comment]
@damemi @ingvagabund I remember we had a conversation about waiting for 1-2 releases before enabling Node Fit filtering by default and removing the redundant checks in the strategies. Is this something we can now implement, or should we wait for a few more releases?

**Comment by damemi** [comment]
I think we should put a release between when we first added it and when we enable it by default, to give enough soak time for users. Does that sound good?

**Comment by RyanDevlin** [comment]
Okay I think that's fair. So essentially once the next release occurs, work can begin on this feature?

**Comment by damemi** [comment]
> Okay I think that's fair. So essentially once the next release occurs, work can begin on this feature?

Yup, that sounds good to me

**Comment by RyanDevlin** [comment]
@damemi I'm going to open up a PR to fix #604 by the end of this week. In the work that was done to implement #604, I ended up consolidating some of the pod filtering logic. Now that it's consolidated, I see a clear path to fixing the issue in this thread as well.

Do you think I should do these in two separate PRs, or should I follow through with aligning all the strategies in my current work?

**Comment by damemi** [comment]
@RyanDevlin I think 2 separate PRs is fine. Refactoring work can always be a follow up

**Comment by RyanDevlin** [comment]
@damemi I didn't want to go against what you said here, but I ended up putting it in [one PR](https://github.com/kubernetes-sigs/descheduler/pull/635). After reviewing the strategies I realized that my work for #604 completed everything but the test cases needed for this issue. Thus, in order to make them two PR's I would have had to undo a bunch of work, and then redo it for the second PR.

So instead I just added the missing test cases and submitted them as one. I hope that's okay, let me know if you'd still like me to break this into two PRs.
